mdp/nodes/sfa_nodes.py

```diff
@@ -145,22 +208,33 @@ def _stop_training(self, debug=False):
 
         rng = self._set_range()
 
+        ####################################################################### EIG
+
         #### solve the generalized eigenvalue problem
         # the eigenvalues are already ordered in ascending order
-        try:
-            self.d, self.sf = self._symeig(self.dcov_mtx, self.cov_mtx,
-                                     range=rng, overwrite=(not debug))
-            d = self.d
-            # check that we get only *positive* eigenvalues
-            if d.min() < 0:
-                err_msg = ("Got negative eigenvalues: %s."
-                           " You may either set output_dim to be smaller,"
-                           " or prepend the SFANode with a PCANode(reduce=True)"
-                           " or PCANode(svd=True)"% str(d))
-                raise NodeException(err_msg)
-        except SymeigException as exception:
-            errstr = str(exception)+"\n Covariance matrices may be singular."
-            raise NodeException(errstr)
+        if not self._use_svd:
+            try:
+                self.d, self.sf = self._symeig(self.dcov_mtx, self.cov_mtx,
+                                         range=rng, overwrite=(not debug))
+                d = self.d
+                # check that we get only *positive* eigenvalues
+                if d.min() < 0:
+                    err_msg = ("Got negative eigenvalues: %s."
+                               " You may either set output_dim to be smaller,"
+                               " or prepend the SFANode with a PCANode(reduce=True)"
+                               " or PCANode(svd=True)"% str(d))
+                    raise NodeException(err_msg)
+
+            except SymeigException as exception:
+                errstr  = str(exception)+"\n Covariance matrices may be singular."
+                errstr += self._data_fail_msg()  # offer some additional information on this issue
+                self._data_analysis()            # and print some information about the training set
```

Comment:
If you want to do some diagnostic, it needs to be outside this class. You cannot mix the computational part and the reporting part, users need to be able to use SFANode without any side effects, and do their own reporting if they want.

For example, people who want to do the data analysis might do
```
try:
   node.stop_training()
except NodeException:
  sfa_diagnostic_report(node)
```

Reply:
One problem we'd like to address is that if `SFANode.stop_training()` crashes then there should be more information provided to help (especially) newcomers understand what the problem is without having to figure it out themselves. The moment it crashes the node should provide what is needed to address the problem.

I can see how encapsulating the diagnostics in an optional method call is cleaner, but I wouldn't see the current solution as problematic? (Though I would agree that the here proposed mechanism is probably not ideal; right now I primarily want *some* solution in the code which we can still improve along the way.)